File: arrays_and_strings/num_subarray_product_less_than_k.py
def num_subarray_product_less_than_k(nums: List[int], k: int) -> int:
    if k <= 1:
        return 0

    num_subarrays = left = 0
    curr_product = 1

    for right in range(len(nums)):
        curr_product *= nums[right]
        
        while curr_product >= k:
            curr_product //= nums[left]
            left += 1

        num_subarrays += right - left + 1
    
    return num_subarrays

File: arrays_and_strings/num_subarray_product_less_than_k.java
public int numSubarrayProductLessThanK(int[] nums, int k)
{
    if (k <= 1)
    {
        return 0;
    }

    int num_subarrays = 0;
    int curr_product = 1;
    int left = 0;

    for (int right = 0; right < nums.length; right++)
    {
        curr_product *= nums[right];
        while (curr_product >= k)
        {
            curr_product /= nums[left++];
        }
        num_subarrays += right - left + 1;
    }

    return num_subarrays;
}